### PR TITLE
[Merged by Bors] - feat: add missing List.foldr lemmas

### PR DESCRIPTION
**Pull request by kim-em**

### CODE REVIEW

File: Mathlib/Data/List/Basic.lean [hunk 1]
```diff
@@ -1373,6 +1373,12 @@ theorem foldl_op_eq_op_foldr_assoc :
   | a :: l, a₁, a₂ => by
     simp only [foldl_cons, foldr_cons, foldl_assoc, ha.assoc]; rw [foldl_op_eq_op_foldr_assoc]
 
+theorem foldr_assoc : ∀ {l : List α} {a₁ a₂}, l.foldr op (op a₁ a₂) = op (l.foldr op a₁) a₂
```
Comment (YaelDillies) [inline Mathlib/Data/List/Basic.lean:1376]:
```suggestion
theorem foldr_left_comm : ∀ {l : List α} {a₁ a₂}, l.foldr op (op a₁ a₂) = op (l.foldr op a₁) a₂
```
no?

Comment (kim-em) [reply Mathlib/Data/List/Basic.lean:1376]: This is consistent with `foldl_assoc` immediately above. Happy if these get renamed together subsequently.

**Comment by YaelDillies** [comment]
maintainer merge

File: Mathlib/Data/List/Perm.lean [hunk 2]
```diff
@@ -51,6 +51,25 @@ theorem Perm.subset_congr_left {l₁ l₂ l₃ : List α} (h : l₁ ~ l₂) : l
 theorem Perm.subset_congr_right {l₁ l₂ l₃ : List α} (h : l₁ ~ l₂) : l₃ ⊆ l₁ ↔ l₃ ⊆ l₂ :=
   ⟨fun h' => h'.trans h.subset, fun h' => h'.trans h.symm.subset⟩
 
+/-- Variant of `Perm.foldr_eq` with explicit commutativity argument. -/
+theorem Perm.foldr_eq' {f : α → β → β} {l₁ l₂ : List α} (p : l₁ ~ l₂)
+    (comm : ∀ x ∈ l₁, ∀ y ∈ l₁, ∀ (z), f y (f x z) = f x (f y z))
+    (init) : foldr f init l₁ = foldr f init l₂ := by
```
Comment (urkud) [inline Mathlib/Data/List/Perm.lean:57]:
```suggestion
    (comm : ∀ x ∈ l₁, ∀ y ∈ l₁, ∀ z, f y (f x z) = f x (f y z))
    (init : β) : foldr f init l₁ = foldr f init l₂ := by
```
Can you use it to prove `Perm.foldr_eq`?

Comment (kim-em) [reply Mathlib/Data/List/Perm.lean:57]: I'd prefer not to scope creep, please.

**Review by Ruben-VandeVelde — COMMENTED** [review]
maintainer merge

**Pull request closed without merging**